mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp

```diff
@@ -0,0 +1,571 @@
+    os << map << "\n";
+  os << "\n";
+}
+
+//===----------------------------------------------------------------------===//
+// ShardingInterface::getShardingOptionFromAttr
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+constexpr StringRef getShardingArrayName() { return "sharding_array"; }
+
+constexpr StringRef getMeshClusterName() { return "mesh_cluster"; }
+
+} // namespace
+
+FailureOr<ShardingOption> mesh::ShardingInterface::getShardingOptionFromAttr() {
+  Operation *op = getOperation();
+  auto arrayAttr = op->getAttrOfType<ArrayAttr>(getShardingArrayName());
+  if (!arrayAttr)
+    return failure();
+  auto symbolRefAttr = op->getAttrOfType<SymbolRefAttr>(getMeshClusterName());
+  if (!symbolRefAttr)
+    return failure();
+  return ShardingOption(getArrayOfI32Array(arrayAttr), symbolRefAttr);
+}
+
+//===----------------------------------------------------------------------===//
+// ShardingInterface::setShardingOptionAttr
+//===----------------------------------------------------------------------===//
+
+void mesh::ShardingInterface::setShardingOptionAttr(
+    Builder &b, const ShardingOption &option) {
+  if (option.empty)
+    return;
+  Operation *op = getOperation();
+  ArrayAttr shardingArrayAttr = b.getArrayOfI32ArrayAttr(option.shardingArray);
+  op->setDiscardableAttr(getMeshClusterName(), option.cluster);
+  op->setDiscardableAttr(getShardingArrayName(), shardingArrayAttr);
+}
+
+//===----------------------------------------------------------------------===//
+// detail::defaultGetShardingOption
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+// Update the given `shardingOption` according to `meshAxes` and `loopIdx`
+static LogicalResult
+fillShardingOption(Operation *op, ShardingOption &shardingOption,
+                   SymbolRefAttr cluster, ArrayRef<int32_t> meshAxes,
+                   unsigned loopIdx, bool ignoreIfConflicted = false) {
+  if ((shardingOption.cluster && cluster &&
+       shardingOption.cluster != cluster) ||
+      (!shardingOption.shardingArray[loopIdx].empty() &&
+       shardingOption.shardingArray[loopIdx] != meshAxes)) {
+    if (ignoreIfConflicted)
+      return success();
+    else
+      return op->emitOpError()
+             << "sharding option confilicts on loop iterator " << loopIdx;
+  }
+  for (size_t i = 0; i < shardingOption.shardingArray.size(); ++i) {
+    if (i != loopIdx) {
+      for (int32_t axis : meshAxes)
+        if (std::find(shardingOption.shardingArray[i].begin(),
+                      shardingOption.shardingArray[i].end(),
+                      axis) != shardingOption.shardingArray[i].end()) {
+          if (ignoreIfConflicted)
+            return success();
+          else
+            return op->emitOpError()
+                   << "sharding option confilicts because of mesh axis " << axis
+                   << " duplicates";
+        }
+    }
```

Comment:
```suggestion
  for (size_t i = 0; i < shardingOption.shardingArray.size(); ++i) {
    if (i == loopIdx) continue;
    for (int32_t axis : meshAxes) {
      if (std::find(shardingOption.shardingArray[i].begin(),
                    shardingOption.shardingArray[i].end(),
                    axis) != shardingOption.shardingArray[i].end()) {
        if (ignoreIfConflicted) {
          return success();
        } else {
          return op->emitOpError()
                 << "sharding option confilicts because of mesh axis " << axis
                 << " duplicates";
        }
      }
    }
```

(nit: early continue)